contracts/DXswapFeeReceiver.sol

```diff
@@ -6,122 +6,226 @@ import './interfaces/IWETH.sol';
+    function _takeTokenOrETH(
+        address pair,
+        address token,
+        uint256 amount
+    ) internal {
+        if (token != WETH && _isSwapPossible(token, WETH, amount)) {
+            // If it is not WETH and there is a direct path to WETH, swap tokens
+            uint256 amountOut = _swapTokensForETH(amount, token);
+            _splitAndTransferFee(pair, WETH, amountOut);
+        } else {
+            // If it is WETH or there is not a direct path from token to WETH, transfer tokens
+            _splitAndTransferFee(pair, token, amount);
+        }
     }
-    
+
     // Take what was charged as protocol fee from the DXswap pair liquidity
     function takeProtocolFee(IDXswapPair[] calldata pairs) external {
-        for (uint i = 0; i < pairs.length; i++) {
+        for (uint256 i = 0; i < pairs.length; i++) {
             address token0 = pairs[i].token0();
             address token1 = pairs[i].token1();
             pairs[i].transfer(address(pairs[i]), pairs[i].balanceOf(address(this)));
-            (uint amount0, uint amount1) = pairs[i].burn(address(this));
-            if (amount0 > 0)
-                _takeETHorToken(token0, amount0);
-            if (amount1 > 0)
-                _takeETHorToken(token1, amount1);
+            (uint256 amount0, uint256 amount1) = pairs[i].burn(address(this));
+            if (amount0 > 0) _takeTokenOrETH(address(pairs[i]), token0, amount0);
+            if (amount1 > 0) _takeTokenOrETH(address(pairs[i]), token1, amount1);
         }
+        emit TakeProtocolFee(msg.sender, ethReceiver, pairs.length);
+    }
+
+    // called by the owner to set maximum swap price impact allowed for single token-weth swap (within 0-100% range)
+    function setMaxSwapPriceImpact(uint32 _maxSwapPriceImpact) external {
+        require(msg.sender == owner, 'DXswapFeeReceiver: CALLER_NOT_OWNER');
+        require(_maxSwapPriceImpact > 0 && _maxSwapPriceImpact < 10000, 'DXswapFeeReceiver: FORBIDDEN_PRICE_IMPACT');
+        maxSwapPriceImpact = _maxSwapPriceImpact;
+    }
+
+    // called by the owner to set external fee receiver address
+    function setExternalFeeReceiver(address _pair, address _externalReceiver) external {
+        require(msg.sender == owner, 'DXswapFeeReceiver: CALLER_NOT_OWNER');
+        externalFeeReceivers[_pair].externalReceiver = _externalReceiver;
     }
 
+    // called by the owner to set fee percentage to external receiver
+    function setFeePercentageToExternalReceiver(address _pair, uint32 _feePercentageToExternalReceiver) external {
+        require(msg.sender == owner, 'DXswapFeeReceiver: CALLER_NOT_OWNER');
+        IDXswapPair swapPair = IDXswapPair(_pair);
+        uint256 feeReceiverBalance = swapPair.balanceOf(address(this));
+        if (feeReceiverBalance > 0) {
+            // withdraw accumulated fees before updating the split percentage
+            address token0 = swapPair.token0();
```

Comment:
This common logic could probably be extracted in a function to avoid duplicated code.